util/shape-options.hh

```diff
@@ -269,6 +269,186 @@ struct shape_options_t
     return ret;
   }
 
+  bool verify_buffer_unsafe_to_concat (hb_buffer_t  *buffer,
+				       hb_buffer_t  *text_buffer,
+				       hb_font_t    *font,
+				       const char  **error=nullptr)
+  {
+    if (cluster_level != HB_BUFFER_CLUSTER_LEVEL_MONOTONE_GRAPHEMES &&
+	cluster_level != HB_BUFFER_CLUSTER_LEVEL_MONOTONE_CHARACTERS)
+    {
+      /* Cannot perform this check without monotone clusters. */
+      return true;
+    }
+
+    /* Check that shuffling up text before shaping at safe-to-concat points
+     * is indeed safe. */
+
+    /* This is what we do:
+     *
+     * 1. We shape text once. Then segment the text at all the safe-to-concat
+     *    points;
+     *
+     * 2. Then we create two buffers, one containing all the even segments and
+     *    one all the odd segments.
+     *
+     * 3. Because all these segments were safe-to-concat at both ends, we
+     *    expect that concatenating them and shaping should NOT change the
+     *    shaping results of each segment.  As such, we expect that after
+     *    shaping the two buffers, we still get cluster boundaries at the
+     *    segment boundaries, and that those all are safe-to-concat points.
+     *    Moreover, that there are NOT any safe-to-concat points within the
+     *    segments.
+     *
+     * 4. Finally, we reconstruct the shaping results of the original text by
+     *    simply interleaving the shaping results of the segments from the two
+     *    buffers, and assert that the total shaping results is the same as
+     *    the one from original buffer in step 1.
```

Comment:
Would it be a useful addition to interleave them shifted as well (if the original is `1odd-2even-3odd-4even`, then interleave them as `2even-1odd-4even-3odd`, and compare original shaping and reconstructed shaping results for that set as well? This creates some additional new "boundary situations" similar to the folding/pulling together in 2.